File: Java/1382.BalanceBinarySearchTree/Solution.java
/* BST: Time:O(n), Space:O(n)
 * 1. Put tree nodes into sortedList
 * 2. Insert the val into the new tree by selecting the mid value of sortedList
 */     

import java.util.*; // Stack

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution {
    private void insertToList(TreeNode root, List<TreeNode> sortedList){
        if(root == null){
            return;
        }
        insertToList(root.left, sortedList);
        sortedList.add(root);
        insertToList(root.right, sortedList);
    }
    
    private TreeNode helper(int lb, int hb, List<TreeNode> sortedList){
        if(lb > hb){
            return null;
        }
        int mid = (lb + hb) / 2;
        TreeNode root = new TreeNode(sortedList.get(mid).val);
        root.left = helper(lb, mid - 1, sortedList);
        root.right = helper(mid + 1, hb, sortedList);
        return root;
    }
    
    public TreeNode balanceBST(TreeNode root) {
        List<TreeNode> sortedList = new ArrayList<>();
        insertToList(root, sortedList);
        int midIdx = (0 + sortedList.size()) / 2;
        return helper(0, sortedList.size() - 1, sortedList);
    }
     
    public static void main(String[] args){
        Solution sol = new Solution();
        TreeNode root = new TreeNode(1);
        root.right = new TreeNode(2);
        root.right.right = new TreeNode(3);
        root.right.right.right = new TreeNode(4);
        TreeNode newRoot = sol.balanceBST(root);
        System.out.println("No example");
    }
}

File: Go/1382.BalanceBinarySearchTree/Solution.go
/* BST: Time:O(n), Space:O(n)
 * 1. Put tree nodes into sortedList
 * 2. Insert the val into the new tree by selecting the mid value of sortedList
 */

package main

import ("fmt")

type TreeNode struct {
    Val int
    Left *TreeNode
    Right *TreeNode
}

func inOrderVisit(root *TreeNode, sortedList *[]*TreeNode){
    if(root == nil){
        return
    }
    inOrderVisit(root.Left, sortedList)
    *sortedList = append(*sortedList, root)
    inOrderVisit(root.Right, sortedList)
}

func getNode(lb int, hb int, sortedList []*TreeNode) *TreeNode {
    if(lb > hb){
        return nil
    }else{
        var mid int = (hb + lb) / 2
        var newNode TreeNode = TreeNode{Val: sortedList[mid].Val}
        newNode.Left = getNode(lb, mid - 1, sortedList)
        newNode.Right = getNode(mid + 1, hb, sortedList)
        return &newNode
    }
}

func balanceBST(root *TreeNode) *TreeNode {
    var sortedList []*TreeNode = make([]*TreeNode, 0, 0)
    inOrderVisit(root, &sortedList)
    return getNode(0, len(sortedList) - 1, sortedList)
}

func main() {
    var root TreeNode = TreeNode{Val:1}
    root.Right = &TreeNode{Val:2}
    root.Right.Right = &TreeNode{Val:3}
    root.Right.Right.Right = &TreeNode{Val:4}
    fmt.Println("No example")
}
